src/rust/engine/process_execution/src/local.rs

```diff
@@ -1,85 +1,132 @@
+  pub fn new(store: fs::Store, fs_pool: Arc<fs::ResettablePool>) -> CommandRunner {
+    CommandRunner { store, fs_pool }
+  }
+
+  ///
+  /// Runs a command on this machine in the passed working directory.
+  ///
+  pub fn run(
+    &self,
+    req: ExecuteProcessRequest,
+    workdir: &Path,
+  ) -> BoxFuture<ExecuteProcessResult, String> {
+    let output = try_future!(
+      Command::new(&req.argv[0])
+        .args(&req.argv[1..])
+        .current_dir(workdir)
+        .env_clear()
+        // It would be really nice not to have to manually set PATH but this is sadly the only way
+        // to stop automatic PATH searching.
+        .env("PATH", "")
+        .envs(req.env)
+        .output().map_err(|e| format!("Error executing process: {:?}", e))
+    );
+
+    let output_snapshot = if req.output_files.is_empty() {
+      future::ok(fs::Snapshot::empty()).to_boxed()
+    } else {
+      // Use no ignore patterns, because we are looking for explicitly listed paths.
+      let posix_fs = Arc::new(fs::PosixFS::new(workdir, self.fs_pool.clone(), vec![]).unwrap());
+      let store = self.store.clone();
+      posix_fs
+        .path_stats(req.output_files.into_iter().collect())
+        .map_err(|e| format!("Error stating output files: {}", e))
+        .and_then(move |paths| {
+          fs::Snapshot::from_path_stats(
+            store.clone(),
+            fs::OneOffStoreFileByDigest::new(store, posix_fs),
+            paths.into_iter().filter_map(|v| v).collect(),
+          )
+        })
+        .to_boxed()
+    };
+
+    output_snapshot
+      .map(|snapshot| ExecuteProcessResult {
+        stdout: Bytes::from(output.stdout),
+        stderr: Bytes::from(output.stderr),
+        exit_code: output.status.code().unwrap(),
+        output_snapshot: snapshot,
+      })
+      .to_boxed()
+  }
 }
 
 #[cfg(test)]
 mod tests {
+  extern crate tempdir;
   extern crate testutil;
 
-  use super::{run_command_locally, ExecuteProcessRequest, ExecuteProcessResult};
   use fs;
-  use std::collections::BTreeMap;
-  use std::path::PathBuf;
+  use futures::Future;
+  use super::{ExecuteProcessRequest, ExecuteProcessResult};
+  use std;
+  use std::collections::{BTreeMap, BTreeSet};
+  use std::env;
+  use std::os::unix::fs::PermissionsExt;
+  use std::path::{Path, PathBuf};
+  use std::sync::Arc;
+  use tempdir::TempDir;
   use self::testutil::{as_bytes, owned_string_vec};
+  use testutil::data::{TestData, TestDirectory};
 
   #[test]
   #[cfg(unix)]
   fn stdout() {
-    let result = run_command_locally(
-      ExecuteProcessRequest {
-        argv: owned_string_vec(&["/bin/echo", "-n", "foo"]),
-        env: BTreeMap::new(),
-        input_files: fs::EMPTY_DIGEST,
-      },
-      &PathBuf::from("/"),
-    );
+    let result = run_command_locally(ExecuteProcessRequest {
+      argv: owned_string_vec(&["/bin/echo", "-n", "foo"]),
+      env: BTreeMap::new(),
+      input_files: fs::EMPTY_DIGEST,
+      output_files: BTreeSet::new(),
```

Comment:
To have similar semantics to globs, this should be either a `Vec` or `OrderMap`... as mentioned elsewhere, I think these should be globs. But if they'll be file paths, we should probably preserve literal order in a similar fashion.

Reply:
Let's come back to this when we've resolved the globs discussion :)

Reply:
Ok, globs resolved, let's talk!

I guess if the `Result` returns a `Directory` this should definitely be a `Set`, but if the `Result` returns a `Snapshot` then either may make sense, and a `Vec` is probably a little more natural.

I'll change this to a `Vec` for now, and see how we go :)